Verify that just-emitted module interfaces parse and typecheck

beccadax opened the pull request:
This PR makes Swift verify module interfaces and private module interfaces after emitting them. To do this, it adds a new frontend mode, `-typecheck-module-from-interface`, and modifies the driver to optionally schedule a typechecking job for each emitted module interface after merge-modules.

~~This is currently done unconditionally, but I'll probably add some kind of flag later on to turn it on only when requested.~~ This extra verification step is enabled by default in assert compilers, but can be enabled in any compiler by passing the `-verify-emitted-module-interface` flag to the driver.

Fixes rdar://66710563.

beccadax committed:
Add a “typecheck module interface” mode

beccadax committed:
Support --driver-mode with swift-frontend

This makes development in Xcode somewhat simpler.

beccadax committed:
Add a debug flag to emit invalid swiftinterface files

beccadax committed:
Tweak diagnostic for failed module interface verification

brentdax commented:
@swift-ci please test

brentdax commented:
I still need to figure out how this should interact with the prebuild script's `-check-only` flag. Should that mode use `-typecheck-from-module-interface` instead of `-compile-from-module-interface`?

nkcsgexi commented on include/swift/Frontend/FrontendOptions.h:
```diff
@@ -116,6 +116,8 @@ class FrontendOptions {
     /// Build from a swiftinterface, as close to `import` as possible
     CompileModuleFromInterface,
+    /// Same as CompileModuleFromInterface, but stopping after typechecking
+    TypecheckModuleFromInterface,
```
Great addition overall! Why not we just reuse `CompileModuleFromInterface` action instead of defining a new one? We could potentially add another flag to indicate stopping after typechecking for `-compile-module-from-interface`. Maintaining an additional frontend action doesn't seem appealing to me.

brentdax replied:
`-typecheck` and `-compile-module-from-interface` are both compiler modes (`ModeOpt`s), so they're mutually exclusive. Also note that their behavior is pretty different—if you take a look at the FrontendOptions.cpp diff, `ActionType::TypecheckModuleFromInterface`'s behavior tends to match `ActionType::Typecheck`, not `ActionType::CompileModuleFromInterface`.

There's something to be said for the idea of removing `-compile-module-from-interface` and instead having people use `-emit-module` to do that; either there would be a flag to restart the compilation in the module interface builder, or this would just happen automatically when you passed a module interface. This would have a couple of nice benefits, like allowing other flags such as `-dump-ast` or `-emit-sil`. But that would be a much bigger change and I'm not sure it can be fully justified.

(Ultimately, I think the new driver probably ought to treat `.swiftinterface` inputs as implicitly being requests to either compiler or typecheck the interface—not sure which. But that's a whole different story, and I think it's better handled in the new driver instead of the old one.)

brentdax commented:
@swift-ci please test

beccadax committed:
Add flag to verify just-emitted module interfaces

The driver can now schedule jobs which typecheck just-emitted module interfaces to ensure that they can be consumed later. This can be enabled manually by passing `-verify-emitted-module-interface` to the driver.

beccadax committed:
Verify interfaces by default in assert compilers

Isolated into its own commit so that it can be reverted easily.

brentdax commented:
SwiftPM test is failing:

```
18:14:35 Test Case '-[CommandsTests.BuildToolTests testParseableInterfaces]' started.
18:14:35 
/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swiftpm/Tests/CommandsTests/BuildToolTests.swift:205: error: -[CommandsTests.BuildToolTests testParseableInterfaces] : failed - error: failed parsing the Swift compiler output: unexpected JSON message: : keyNotFound(CodingKeys(stringValue: "outputs", intValue: nil), Swift.DecodingError.Context(codingPath: [], debugDescription: "No value associated with key CodingKeys(stringValue: \"outputs\", intValue: nil) (\"outputs\").", underlyingError: nil))
```

Seems totally plausible that something might need to be updated there.

brentdax commented:
With apple/swift-package-manager#2854

@swift-ci please test

brentdax commented:
@swift-ci please test
